ssl/record/rec_layer_s3.c

```diff
@@ -1535,33 +1538,37 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
                 return -1;
             }
 
-            if (alert_descr == SSL_AD_CLOSE_NOTIFY) {
-                s->shutdown |= SSL_RECEIVED_SHUTDOWN;
-                return 0;
-            }
             /*
              * Apart from close_notify the only other warning alert in TLSv1.3
              * is user_cancelled - which we just ignore.
              */
-            if (SSL_IS_TLS13(s) && alert_descr != SSL_AD_USER_CANCELLED) {
-                SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER, SSL_F_SSL3_READ_BYTES,
-                         SSL_R_UNKNOWN_ALERT_TYPE);
-                return -1;
-            }
-            /*
-             * This is a warning but we receive it if we requested
-             * renegotiation and the peer denied it. Terminate with a fatal
-             * alert because if application tried to renegotiate it
-             * presumably had a good reason and expects it to succeed. In
-             * future we might have a renegotiation where we don't care if
-             * the peer refused it where we carry on.
-             */
-            if (alert_descr == SSL_AD_NO_RENEGOTIATION) {
-                SSLfatal(s, SSL_AD_HANDSHAKE_FAILURE, SSL_F_SSL3_READ_BYTES,
-                         SSL_R_NO_RENEGOTIATION);
-                return -1;
-            }
-        } else if (alert_level == SSL3_AL_FATAL) {
+            if (tls13_user_cancelled)
+                goto start;
+        }
+
+        if (alert_descr == SSL_AD_CLOSE_NOTIFY
+                && (SSL_IS_TLS13(s) || alert_level == SSL3_AL_WARNING)) {
+            s->shutdown |= SSL_RECEIVED_SHUTDOWN;
+            return 0;
+        }
+
```

Comment:
Now when it's remodeled like that I'd say that what earlier mentioned side-note problem is actually more acute. I mean the lack of else's in a sequence that you expect to be switch-like. The situation is intensified by the fact that sequence now ends with if-else. I wonder what would you think of following. First consider removing `if (tls13_user_cancelled) goto start` from previous if (so that it becomes exclusively about setting cap on amount of warnings, explicit or implicit like with user_cancelled), and then start separate if-chain with it, reordering the rest as following:
```
if (tls13_user_cancelled) {
    goto start;
} else if (alert_descr == SSL_AD_CLOSE_NOTIFY
           && (SSL_IS_TLS13(s) || alert_level == SSL3_AL_WARNING) {
    ...
} else if (alert_level == SSL3_AL_FATAL || SSL_IS_TLS13(s)) {
    ...
} else if (alert_descr == SSL_AD_NO_RENEGOTIATION) {
    /* no need to check for TLS 1.3 or SSL3_AL_FATAL */
    ...
} else {
    ...
}
```
Original comment about tls13_user_cancelled was made in specific context (of nested if's readability), and now, after remodeling, it might not make as much sense as it originally did. Caching the result of SSL_IS_TLS13(s) in is_tls13 might be more sensible. [One can even wonder if it would be appropriate to force alert_level to SSL3_AL_WARNING if is_tls13 is true, to denote that it's effectively ignored. This would allow to simplify logic even further.]